packages/js/data/src/onboarding/with-onboarding-hydration.tsx

```diff
@@ -21,39 +20,54 @@ export const withOnboardingHydration = ( data: {
 		( OriginalComponent ) => ( props ) => {
 			const onboardingRef = useRef( data );
 
-			useSelect(
-				// @ts-expect-error // @ts-expect-error registry is not defined in the wp.data typings
-				( select: ( s: string ) => OnboardingSelector, registry ) => {
-					if ( ! onboardingRef.current ) {
-						return;
-					}
-
+			const { isResolvingGroup, hasFinishedResolutionGroup } = useSelect(
+				( select ) => {
 					const { isResolving, hasFinishedResolution } =
 						select( STORE_NAME );
-					const {
-						startResolution,
-						finishResolution,
-						setProfileItems,
-					} = registry.dispatch( STORE_NAME );
-
-					const { profileItems } = onboardingRef.current;
-
-					if (
-						profileItems &&
-						! hydratedProfileItems &&
-						! isResolving( 'getProfileItems', [] ) &&
-						! hasFinishedResolution( 'getProfileItems', [] )
-					) {
-						startResolution( 'getProfileItems', [] );
-						setProfileItems( profileItems, true );
-						finishResolution( 'getProfileItems', [] );
-
-						hydratedProfileItems = true;
```

Comment:
This is a really weird use of `useSelect` and I'm starting to see how https://github.com/WordPress/gutenberg/pull/47243 could break it. The main issue is that the `useSelect` callback is run _reentrantly_, as the dispatches modify the same store, even the same value, as is being selected.

The sequence of events is:
1. `useSelect` runs the select callback.
2. The callback dispatches the `startResolution` action.
3. After the `startResolution` action modifies the store, it runs all registered listeners.
4. One of the listeners is our own `useSelect` hook.
5. The listener calls the select callback to determine the new selected value.
6. There is a second call of the select callback, at the time when the first call is still running!

Now note that `useSelect` runs the select callback inside the [`__unstableMarkListeningStores`](https://github.com/WordPress/gutenberg/blob/79623c991ca8b37e5b104264b8be12c5ae72e82d/packages/data/src/registry.js#L120-L128) wrapper, which sets and unsets the global `listenerStores` variable. While the first call is still running, the second call resets the shared `listenerStores` set with a new, empty instance.

The `listenerStores` set keeps a list of stores to subscribe to, to watch changes in them. Now, instead of subscribing to `STORE_NAME`, the `useSelect` hook won't subscribe to any stores, because the set is empty, and won't see any store updates, like changes in resolution status.

Before https://github.com/WordPress/gutenberg/pull/49577, `useSelect` used the `__unstableMarkListeningStores` wrapper only on mount, to figure out the set of stores to subscribe to, but after that PR, it uses the wrapper for every select, because it can newly add new subscriptions dynamically. Therefore, there are many more opportunities to hit that reentrancy issue.